ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/FilterTypeValidator.java

```diff
@@ -0,0 +1,84 @@
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.analyzer;
+
+import static java.util.Objects.requireNonNull;
+
+import io.confluent.ksql.engine.rewrite.StatementRewriteForMagicPseudoTimestamp;
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.util.ExpressionTypeManager;
+import io.confluent.ksql.function.FunctionRegistry;
+import io.confluent.ksql.schema.ksql.FormatOptions;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.KsqlStatementException;
+
+/**
+ * Validates types used in filtering statements.
+ */
+public final class FilterTypeValidator {
+
+  private final LogicalSchema schema;
+  private final FunctionRegistry functionRegistry;
+  private final FilterType filterType;
+
+  public FilterTypeValidator(
+      final LogicalSchema schema,
+      final FunctionRegistry functionRegistry,
+      final FilterType filterType
+  ) {
+    this.schema = requireNonNull(schema, "schema");
+    this.functionRegistry = requireNonNull(functionRegistry, "functionRegistry");
+    this.filterType = requireNonNull(filterType, "filterType");
+  }
+
+  /**
+   * Validates the given filter expression.
+   */
+  public void validateFilterExpression(final Expression exp) {
+    // Construct an expression type manager to be used when checking types.
+    final ExpressionTypeManager expressionTypeManager = new ExpressionTypeManager(schema,
+        functionRegistry);
+
+    // Rewrite the expression with magic timestamps, so type checking can pass
+    final Expression magicTimestampRewrite =
+        new StatementRewriteForMagicPseudoTimestamp().rewrite(exp);
+
+    final SqlType type;
+    try {
+      type = expressionTypeManager.getExpressionSqlType(magicTimestampRewrite);
+    } catch (KsqlStatementException e) {
+      throw new KsqlStatementException("Error in " + filterType.name() + " expression: "
+          + e.getRawMessage(), e.getSqlStatement());
+    } catch (KsqlException e) {
+      throw new KsqlStatementException("Error in " + filterType.name() + " expression: "
+          + e.getMessage(), exp.toString());
+    }
+    if (!SqlTypes.BOOLEAN.equals(type)) {
+      throw new KsqlStatementException("Type error in " + filterType.name() + " expression: "
+          + "Should evaluate to boolean but is " + exp.toString()
+          + " (" + type.toString(FormatOptions.none()) + ") instead.",
+          exp.toString());
+    }
```

Comment:
This is a misuse of `KsqlStatementException`.  It takes the _full_ sql statement as its second parameter, not the snippet of the statement that causes the error.

As above, just throw `KsqlException`. Code higher up will handle the conversion to `KsqlStatementException`.